syntax = "proto3";

option go_package = "github.com/d-sparks/gravy/supervisor/proto;supervisor_pb";

import "google/protobuf/timestamp.proto";

package supervisor;

service Supervisor {
  rpc PlaceOrder(Order) returns (OrderConfirmation) {}

  rpc GetPortfolio(AlgorithmId) returns (Portfolio) {}

  rpc SynchronousDailySim(SynchronousDailySimInput) returns (SynchronousDailySimOutput) {}

  rpc Abort(AbortInput) returns (AbortOutput) {}
}

message Order {
  string requester_id = 1;
  string ticker = 2;
  int32 volume = 3;
  double limit = 4;
  double stop = 5;
}

message OrderConfirmation {
}

message Portfolio {
  message Stock {
    string ticker = 1;
    double volume = 2;
  }

  repeated Stock stocks = 1;
  double usd = 2;
}

message AlgorithmId {
  string algorithm_id = 1;
}

message SynchronousDailySimInput {
  google.protobuf.Timestamp start = 1;
  google.protobuf.Timestamp end = 2;
  repeated AlgorithmId algorithms = 3;
}

message SynchronousDailySimOutput {
}

message AbortInput {
}

message AbortOutput {
}